### PR TITLE
Set default max motor current to 2048mA

### PR DESCRIPTION
**Pull request by popy2k14**
An chip erase is necessary that the new value is used!
Also increased version to v83.

### CODE REVIEW

**Comment by popy2k14** [comment]
Could you please compile a binary with the changes so i can test them?
Sadly my compiled binarys are way smaler than yours. So dont know if i am missing something in the build.

Thanks
pOpY

**Comment by popy2k14** [comment]
Sadly, my own build binary does'nt work as @mjuhanne v082_sleep
I have found another way to change various settings (those ones which will be stored as settings in flash).

You can dump the firmware after setting it up with STM32CubeProgrammer or STLink and edit the binary file with an HEX editor like HxD (recommend this one).

!! Attention !! Backup your current firmware and settings as described on the github readme.

The settings are stored @ address 0x7800 in the binary file (flash).
You can search for the last occurence of each setting identifier and modify it's 16bit value (2 bytes) which is stored BEFORE the identifier.

!! Attention !! The value is stored in little endian where the LSB is stored first!

The identifiers for the settings are:

	MAX_CURTAIN_LEN_EEPROM = 0x5555,
	FULL_CURTAIN_LEN_EEPROM = 0x6666,
	MINIMUM_VOLTAGE_EEPROM = 0x7777,
	DEFAULT_SPEED_EEPROM = 0x8888,
	AUTO_CAL_EEPROM = 0x9999,
	ORIENTATION_EEPROM = 0xAAAA,
	MAX_MOTOR_CURRENT_EEPROM = 0xBBBB,
	STALL_DETECTION_TIMEOUT_EEPROM = 0xCCCC

In my case i have searched the last 0xBBBB value and modified the two bytes before from 00 04 (= 1024 dec) to 00 08 (2048 dec).

The best way to find the correct identifier is to go to offset 0x7800. There you can find a bunch bytes of data whichj are not FF. Go to the part where data is filled with subsequents FF and set curesor anywhere in the FFs. Use Search -> HEX -> Reverse and search in reverse for your specified identifier.

After that save your changes and reflash the file.

Hope this helps someone.

**Comment by popy2k14** [comment]
You can also disable the change of next flexi-speed on UP presses.
In my case (deconz & conbee II gateway) or my wife/son :-) has changed the speed by just pressing UP on the ikea switch.
To prevent this i NOP'ed out an if -> results in disabling flexi-speed.

The offset is 26E2 changing 01 D0 to 00 BF.

Firmware is running as expected but with disabled flexi-speed :-)

**Comment by fhempy** [comment]
Hi,
I managed to compile a functional binary. I used VSCode with the STM32 extension on Windows.
I flashed it with the st-flash utility and it works. Changed max current, default speed and enabled sleep mode in the source code and it works.

**Comment by popy2k14** [comment]
nice you got it running.
I am away from the firmware because of too much troubles.
Flexi speed has changed itself and so on.
Also one device was just dead.

**Comment by fhempy** [comment]
I had the flexi speed change issues as well. I'll test the compiled firmware and see if it works there.

I also added a change so that only 2x CMD_UP causes a calibration, otherwise it's too loud when calibrating every morning on CMD_UP.

You can find my code here:
https://github.com/dominikkarall/fyrtur-motor-board

I think there is a bug in the code which causes the need of a stop command sometimes sent before open or close.

**Comment by popy2k14** [comment]
thx for keeping the project alive.
But i am away from modding my blinds because too less time.

**Comment by mjuhanne** [comment]
Hello all, sorry for my inactivity in this project for these past few months.

I've now disabled the current sensing and flexi speed by default, but you can find the precompiled binary files which have flexi speed enabled as well as few preset speeds.   You don't need to erase the flash because I've changed the maximum current EEPROM address so old setting is not used. Also the precompiled fixed speed binaries ignore speed values in EEPROM so you can switch between those without having to erase memory completely.

Here's some more detailed explanation (the same as in README.md):

TLDR; This should be safe BUT DON'T COME AFTER ME WITH A LAWYER IF YOUR HOUSE BURNS DOWN :)

Versions 0.82 (and earlier) have a predefined maximum motor current limit setting (1A default, configurable up to 2A). This software limit can cause annoyances on some motor units which have higher static and/or dynamic friction than anticipated. This causes inconsistently high power draw and thus the firmware makes false assumption that the motor is stalled and the firmware then resets its known position to zero (upmost position) erroneously. When this position mismatch occurs it's a recipe for disaster because when the user (or most likely the home automation system as scheduled) tries to roll blinds down it causes movement further below the absolute lower limit, after which blinds start to move up but in reverse orientation. This can create an annoying mess with wrinkled curtains and even damage the gluing between the curtains and the axel rod!

This false stall detection behaviour has sometimes occured even after blinds have started to rotate a bit and they should be free of static friction, so it's possible that abnormal current spikes might be detected by the current sensing circuitry also after initial movement.

The current limit feature in this firmware was designed to be an extra layer of safety in addition to the fuse on the custom (NOT ORIGINAL!) Fyrtur board. The main method of stall detection is via timeout of signals from the Hall sensors. This is also the way the original firmware works as far as I know. In my reverse engineering project of the original Ikea motor firmware I didn't find any evidence that the current sensing circuitry would be used for stall detection (take this with a grain of salt - it's possible that I have overlooked this part of the code). So it seems that even Ikea (or the subcontractor that manufactured the motor module) didn't bother to use the current sensing circuitry, possibly because of these glitches.

This makes me conclude that in order to avoid annoying false stall detection and position mismatches it should be safe to disable the current sensing feature and to rely solely on the Hall sensor timeout (296 ms by default). After all, the current sensing is an extra feature not implemented originally by the manufacturer. Needless to say, this kind of software feature, even if enabled, doesn't prevent from possible programming errors or physical catastrofic failures. Thus it's mandatory that other precautions are made to prevent too high current consumption in case of actual faulty operation. Most (if not all) power adapters shut down the output if their maximum current draw is exceeded. The original Fyrtur board doesn't have any kind of fuse, but the Li-Ion battery modules are specified to have maximum 2A output. So I would like to assume that they have an internal current limiting / protection circuitry, BUT I HAVEN'T TESTED THEIR SAFETY WHEN SHORT CIRCUITED NOR DO I WANT TO!

So, from version 0.83 and onwards the current sensing feature is disabled, but can be re-enabled if needed via CMD_EXT_SET_MAX_MOTOR_CURRENT command.

**Comment by mjuhanne** [comment]
> I had the flexi speed change issues as well. I'll test the compiled firmware and see if it works there.
> 
> I also added a change so that only 2x CMD_UP causes a calibration, otherwise it's too loud when calibrating every morning on CMD_UP.

In my system I use the GOTO POSITION (0-100) instead of UP command when it's a scheduled movement (like in the morning).
> 
> You can find my code here: https://github.com/dominikkarall/fyrtur-motor-board
> 
> I think there is a bug in the code which causes the need of a stop command sometimes sent before open or close.

Interesting. Let me know if you know how to reproduce this

**Comment by fhempy** [comment]
Hi @mjuhanne 
Great to see you back

Unfortunately my motor control doesn't work anymore. Battery voltage drops to 1V (from 8,2V when only ZigBee module connected) as soon as I connect the controller (motor disconnected). I assume it might be a faulty capacitor. So maybe the issue with the required stop command was already caused by faulty hardware as well. If you have any ideas which component might cause that issue, please let me known. Thanks!

**Comment by mjuhanne** [comment]
@dominikkarall I'm sorry to hear that. Let's hope that it's just a faulty hardware for reasons not related to the custom firmware. I really wouldn't like people's motor controllers to get fried because of this mod. The only way I could see this happening programmatically if somehow the MOSFETs were turned on in wrong combination (both HIGH1 and LOW1 on or HIGH2 and LOW2 on), which would result in blowing up the MOSFETs (they usually fail resulting in short circuit).  I really tried to avoid this by always stopping all PWM and MOSFET drivers first (see motor_start_common() and motor_stop()) and having a short delay before starting the motors.   That being said, I have been running my 5 Fyrturs with CFW for almost 1.5 years now without any failures.

If you really want to dig into this and you're hardware savvy, there's a Kicad schematic for the motor board in this repository. I would first check for a short circuit between +7.5V and GND. However if that occurs then it's a nightmare to try to find where the actual damage is without desoldering each possible component one by one (unless you see something that is clearly damaged or fried).   You could also remove the DC motor itself and make sure the it has enough resistance. If somehow it's windings are damaged and there's short circuit inside the motor, it could blow up the MOSFETs as a result as well.

**Comment by fhempy** [comment]
@mjuhanne thx for pointing me to the MOSFETs (IRF5851). I'll try to exchange those and see if I can get it working again.

It could be that I caused that issue on my own by doing some code changes and experimenting with it.
